path_dynamic_creds.go

```diff
@@ -0,0 +1,242 @@
+package openldap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/go-ldap/ldif"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/base62"
+	"github.com/hashicorp/vault/sdk/helper/template"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func (b *backend) pathDynamicCredsRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("name").(string)
+
+	// Get the role and LDAP configs
+	dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve dynamic role: %w", err)
+	}
+	if dRole == nil {
+		return nil, nil
+	}
+
+	config, err := readConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return nil, fmt.Errorf("missing OpenLDAP configuration")
+	}
+
+	// Generate dynamic data
+	username, err := generateUsername(req, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate username: %w", err)
+	}
+	password, err := b.GeneratePassword(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the template
+	now := time.Now()
+	exp := now.Add(dRole.DefaultTTL)
+	templateData := dynamicTemplateData{
+		Username:              username,
+		Password:              password,
+		DisplayName:           req.DisplayName,
+		RoleName:              roleName,
+		IssueTime:             now.Format(time.RFC3339),
+		IssueTimeSeconds:      now.Unix(),
+		ExpirationTime:        exp.Format(time.RFC3339),
+		ExpirationTimeSeconds: exp.Unix(),
+	}
+	createLDIF, err := applyTemplate(dRole.CreationLDIF, templateData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply creation_ldif template: %w", err)
+	}
+
+	// Parse the raw LDIF & run it against the LDAP client
+	entries, err := ldif.Parse(createLDIF)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse generated LDIF: %w", err)
+	}
+
+	addReq := getAddRequest(entries.Entries[0])
+
+	err = b.client.Add(config.LDAP, addReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create LDAP entry: %w", err)
+	}
+	respData := map[string]interface{}{
+		"username": username,
+		"password": password,
+		"dn":       addReq.DN,
+	}
+	internal := map[string]interface{}{
+		"name":     roleName,
+		"username": username,
+		"password": password,
```

Comment:
Do we need to store the username and password as internal data? AFAIK the data in this map is used for renewal and revocation, but the handlers for those shouldn't need `username` and `password`.

Reply:
This has been changed to include the `deletion_ldif` as well as all of the dynamic template data so the deletion template can receive all of the same data the creation template does.

Reply:
I'm not quite getting your comments. AFAIK the internal data map is only really useful for lease-related logic (i.e. renew/revoke), but I don't see those entries being used on `secretCredsRenew` or `secretCredsRevoke` when we're accessing the values in this map.

Reply:
`username` and `password` are not top level fields any longer. `deletion_ldif` is being set in the internal data and used during `secretCredsRevoke` along with `template_data` which contains all of the same information that is applied to the `creation_ldif` template during credential creation.

Reply:
Argh, you're right on both counts -- I was looking at an outdated diff in my other tab.